core/services/relay/evm/chain_reader.go

```diff
@@ -261,6 +357,68 @@ func (cr *chainReader) addDecoderDef(contractName, methodName string, outputs ab
 	return output.Init()
 }
 
+func remapQueryKeyFilters(key string, eventIndexBindings EventIndexBindings, queryFilter commontypes.QueryFilter) (commontypes.QueryFilter, error) {
+	var logFilters []commontypes.Expression
+	for _, expression := range queryFilter.Expressions {
+		logFilter, err := remapExpression(expression)
+		if err != nil {
+			return commontypes.QueryFilter{}, err
+		}
+
+		logFilters = append(logFilters, logFilter)
+	}
+
+	eventSig, address, _, err := eventIndexBindings.Get(key)
+	if err != nil {
+		return commontypes.QueryFilter{}, err
+	}
+
+	queryFilter.Expressions = append(logFilters, NewEventFilter(address, eventSig))
+
+	return queryFilter, nil
+}
+
+// TODO values can't be string?
+func remapQueryKeyByValuesFilters(keyValueIn commontypes.KeyValueIn, eventIndexBindings EventIndexBindings, queryFilter commontypes.QueryFilter) (commontypes.QueryFilter, error) {
+	eventSig, address, topicIndex, err := eventIndexBindings.Get(keyValueIn.Key)
```

Comment:
I'm still trying to understand what the purpose of `QueryKey()` & `QueryKeys()` is, and it seems like this might be the key here but I'm pretty confused.

I thought the generic key names were supposed to be the topic index, for example if a contract emits an event that looks like:
`emit Transfer(from, to, amount)`
the generic keys should be:  `EventType`, `from`, `to`, and `amount` and then the values they can take are things like "Transfer", "address1", "address2", and 5000.

But here you're taking one of the generic key names, and using it to look up an event sig, an address, and a topicIndex.  The topic index is a chain-specific key so that one makes sense, but eventSig is a value not a key... and address is neither.  Seems like I must be missing something?

Reply:
If you look [where the key mappings are set](https://github.com/smartcontractkit/chainlink/blob/c466c600ae0d5c2b9379f513e9460e69355ababc/core/services/relay/evm/chain_reader.go#L285), you can see that keys represent two different things.
1. First is just a general pointer to a type of data which in EVM case is eventSignature. From user perspective this means using `QueryKey()` to just say give me all sequences which match this type of key
2. Second is a pointer to a value inside of a type of data which in EVM case means pointing to a specific event topic or data word byte. From user perspective this means using `QueryByKeyValuesComparison` to match specific values.

Reply:
Ah, ok... I think I get the general idea. I had assumed we would just be treating EventSig like any other key but seems that we treat it as its own special thing. I was assuming we'd want to not have it treated special in case some blockchains don't have any analogue of it. But I guess all of our products have some notion of an "event type" so, it could make sense from a user interface perspective. I'll have to look at it a bit more to get a feel for how it will work in practice.

Reply:
On a somewhat relevant side note: one of the changes to LogPoller we'll need to make if we decide on going forward with supporting Tron is passing the string representation of each event type to LogPoller instead of the EventSig.  Tron needs the string representation of the event name (without any parenthesis or arguments) sent directly to the rpc server instead of sending a hash of the whole signature.  I'm not sure what happens on Tron if you have two events named the same thing but different types, like `ValueReceived(uint64 val)` and `ValueReceived(string val)`... on evm these would both have the same name but different event sigs while on Tron the equivalent key-specific values would be identical. (I'm guessing they just don't allow that.)

Reply:
Re to log poller changes. 
Won't the chain reader be able to infer that from the config? We just bind to an "event" and the event details are in the config

Reply:
Re to how eventSig is treated.

I am not looking at it from the perspective of event signature. I am just treating it as we have keys that point to an idea of something and keys that point to a filter over that concept.

give me sequences for  "transfer" or give me sequences where "transfer-value" is 1000, both of these just return sequences(events) of transfer in the end

Reply:
If you try to say give me sequences of "transfer" where value is 1000 it would return that the key is invalid as that key is not searchable by value. If you say give me sequences of "transfer-from"  we can either return all transfer sequences or treat it as an invalid value also